components/base/sensorcontrolled/movestraight.go

```diff
@@ -0,0 +1,204 @@
+// Package sensorcontrolled base implements a base with feedback control from a movement sensor
+package sensorcontrolled
+
+import (
+	"context"
+	"errors"
+	"math"
+	"time"
+
+	geo "github.com/kellydunn/golang-geo"
+
+	rdkutils "go.viam.com/rdk/utils"
+)
+
+const (
+	slowDownDistGain      = .1
+	maxSlowDownDist       = 100 // mm
+	moveStraightErrTarget = 20  // mm
+	headingGain           = 1.
+)
+
+func (sb *sensorBase) MoveStraight(
+	ctx context.Context, distanceMm int, mmPerSec float64, extra map[string]interface{},
+) error {
+	ctx, done := sb.opMgr.New(ctx)
+	defer done()
+	sb.setPolling(false)
+	straightTimeEst := time.Duration(int(time.Second) * int(math.Abs(float64(distanceMm)/mmPerSec)))
+	startTime := time.Now()
+	timeOut := 5 * straightTimeEst
+	if timeOut < 10*time.Second {
+		timeOut = 10 * time.Second
+	}
+
+	if sb.position == nil || len(sb.controlLoopConfig.Blocks) == 0 {
+		sb.logger.CWarnf(ctx,
+			"Position reporting sensor not available, and no control loop is configured, using base %s MoveStraight",
+			sb.controlledBase.Name().ShortName())
+		sb.stopLoop()
+		return sb.controlledBase.MoveStraight(ctx, distanceMm, mmPerSec, extra)
+	}
+
+	initialHeading, err := sb.headingFunc(ctx)
+	if err != nil {
+		return err
+	}
+	// make sure the control loop is enabled
+	if sb.loop == nil {
+		if err := sb.startControlLoop(); err != nil {
+			return err
+		}
+	}
+
+	// initialize relevant parameters for moving straight
+	slowDownDist := calcSlowDownDist(distanceMm)
+
+	var initPos *geo.Point
+
+	if sb.position != nil {
+		initPos, _, err = sb.position.Position(ctx, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	ticker := time.NewTicker(time.Duration(1000./sb.controlLoopConfig.Frequency) * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			// do not return context canceled errors, just log them
+			if errors.Is(ctx.Err(), context.Canceled) {
+				sb.logger.Error(ctx.Err())
+				return nil
+			}
```

Comment:
switched to logging context canceled errors cc @randhid

Reply:
so if you're using the API context in your busy loop, you don't know where that is coming from and what is cancelling it. If you switch to a background context spawned from within this code that you have full control lover, that is the preferred way to do things.

We can follow up context clean up as we test with the grid pattern and sine velocities.